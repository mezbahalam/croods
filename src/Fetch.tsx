import React, { useEffect } from 'react'

import { useCroods } from './useCroods'
import { useBaseOptions } from './baseOptionsProvider'

import type { UseCroodsOptions, CroodsTuple } from './useCroods'
import { CroodsState } from './types'

type Unpack<T> = T extends Array<any> ? T[number] : T
type InfoOrList<T> = T extends Array<any>
  ? CroodsState<T[number]>['list']
  : CroodsState<T>['info']
type FetchOptions<T> = Omit<UseCroodsOptions, 'fetchOnMount'> & {
  render: (t: InfoOrList<T>, b: CroodsTuple<Unpack<T>>) => React.ReactNode
}

function Fetch<T = any>({
  id,
  query,
  path,
  stateId,
  render,
  ...opts
}: FetchOptions<T>): JSX.Element {
  const baseOptions = useBaseOptions()
  const options: UseCroodsOptions = {
    ...baseOptions,
    ...opts,
    id,
    path,
    stateId,
  }
  const [state, actions] = useCroods<Unpack<T>>(options)
  const error = state.listError || state.infoError
  const isList = !id
  const result = isList ? state.list : state.info

  useEffect(() => {
    actions.fetch({ id })(query)
  }, [id, query, path, stateId])

  if (isList ? state.fetchingList : state.fetchingInfo) {
    return <>{options.renderLoading?.() || <div>Loading...</div>}</>
  }

  if (error) {
    return (
      <>
        {options.renderError?.(error) || (
          <div style={{ color: 'red' }}>{error}</div>
        )}
      </>
    )
  }

  if (!isList && !state.info && options.renderEmpty) {
    return <>{options.renderEmpty?.() || null}</>
  }

  if (isList && !Boolean(state.list?.length) && options.renderEmpty) {
    return <>{options.renderEmpty() || null}</>
  }

  return <>{render(result as InfoOrList<T>, [state, actions])}</>
}

export { Fetch }
export type { FetchOptions }
