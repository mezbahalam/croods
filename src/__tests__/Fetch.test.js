import React from 'react'
import renderer from 'react-test-renderer'
import { render } from '@testing-library/react'

import Fetch from '../Fetch'

let mockState = {}
const mockActions = {
  fetch: jest.fn(() => jest.fn()),
}

afterEach(() => {
  jest.clearAllMocks()
})

jest.mock('../useCroods', () => () => [mockState, mockActions])

const props = {
  id: 1,
  name: 'foobar',
  renderEmpty: () => <div>Empty</div>,
  render: () => <div>Rendered</div>,
}

describe('with error', () => {
  it('renders correctly', () => {
    mockState = {
      infoError: 'Error foobar',
    }
    const tree = renderer.create(<Fetch {...props} />).toJSON()
    expect(tree).toMatchSnapshot()
  })
})

describe('while is loading', () => {
  it('renders correctly', () => {
    mockState = {
      fetchingInfo: true,
    }
    const tree = renderer.create(<Fetch {...props} />).toJSON()
    expect(tree.children[0]).toBe('Loading...')
  })

  it('does not render loading if fetching list and showing info', () => {
    mockState = {
      fetchingList: true,
    }
    const tree = renderer.create(<Fetch {...props} />).toJSON()
    expect(tree.children[0]).not.toBe('Loading...')
  })

  it('does not render loading if fetching info and showing list', () => {
    mockState = {
      fetchingInfo: true,
    }
    const tree = renderer.create(<Fetch {...props} id={undefined} />).toJSON()
    expect(tree.children[0]).not.toBe('Loading...')
  })
})

describe('with info', () => {
  it('renders correctly', () => {
    mockState = {
      info: {
        id: 1,
        data: {
          foo: 'bar',
        },
      },
    }
    const tree = renderer.create(<Fetch {...props} />).toJSON()
    expect(tree).toMatchSnapshot()
  })
})

describe('when info is empty', () => {
  it('renders correctly', () => {
    mockState = {
      info: null,
    }
    const tree = renderer.create(<Fetch {...props} />).toJSON()
    expect(tree).toMatchSnapshot()
  })
})

describe('fetching list', () => {
  const fetchListProps = { ...props, id: undefined }
  describe('when list is empty', () => {
    it('renders correctly', () => {
      mockState = {
        info: null,
        list: [],
      }
      const tree = renderer.create(<Fetch {...fetchListProps} />).toJSON()
      expect(tree).toMatchSnapshot()
    })
  })
})

describe('when changing properties', () => {
  it('should only fetch once if no property is changed', () => {
    const { rerender } = render(<Fetch {...props} />)
    rerender(<Fetch {...props} foo="bar" />)
    expect(mockActions.fetch).toHaveBeenCalledTimes(1)
  })

  it('should fetch when changing stateid', () => {
    const { rerender } = render(<Fetch {...props} />)
    rerender(<Fetch {...props} stateId="foobar" />)
    rerender(<Fetch {...props} query={{ foo: 'bar' }} stateId="foobar" />)
    rerender(
      <Fetch
        {...props}
        path="/foobar"
        query={{ foo: 'bar' }}
        stateId="foobar"
      />,
    )
    expect(mockActions.fetch).toHaveBeenCalledTimes(4)
  })
})
