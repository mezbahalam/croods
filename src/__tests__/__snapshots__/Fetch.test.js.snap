// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fetching list when list is empty renders correctly 1`] = `
<div>
  Empty
</div>
`;

exports[`when info is empty renders correctly 1`] = `
<div>
  Empty
</div>
`;

exports[`with error renders correctly 1`] = `
<div
  style={
    Object {
      "color": "red",
    }
  }
>
  Error foobar
</div>
`;

exports[`with info renders correctly 1`] = `
<div>
  Rendered
</div>
`;
